### PR TITLE
Add codec versioning support

### PR DESCRIPTION
**Pull request by mdumandag**
This PR adds the ability to add new parameters, methods, custom types and events to the client protocol. 

As discussed within the team, existence (or lack of it) of the new parameters can be checked by the `isXXExists` fields/arguments of the RequestParameters, ResponseParameters, handleYYEvent methods, custom type constructor calls/factory method calls.

Also, for the custom types, the following change is made. If the custom type does not have an initial frame before (no fix sized paramters), and we extend the custom type with new fix sized parameters, those parameters are encoded into the `BEGIN_FRAME` of the custom type since adding a new initial frame would break the old readers.

### CODE REVIEW

File: schema/protocol-schema.json [hunk 1]
```diff
@@ -6,7 +6,8 @@
   "definitions": {
     "since": {
       "enum": [
-        2.0
+        2.0,
+        2.1
```
Comment (mdumandag) [inline schema/protocol-schema.json]: Since we will have changes in the 2.0.1 branch, should we add it here ?

File: README.md [hunk 2]
```diff
@@ -220,4 +219,29 @@ enum type to int is required. In the factory method, you will receive an integer
...
+Client protocol can be expanded by adding new
+* services
+* methods
+* parameters to existing method requests, responses or events
+* events to existing methods
+* custom types
+* parameters to existing custom types
+
+While expanding the protocol, one needs to follow these simple guidelines:
+* `since` field of the protocol definitions of the the newly added parameters, methods, events and custom types should
```
Comment (mdumandag) [inline README.md]: Where should we put the current protocol version ? Should it be on README, generator.py or somewhere else ?

Comment (asimarslan) [reply README.md]: The latest since tag will define the current version. Maybe a log during code generation will be helpful.

Comment (asimarslan) [reply README.md]: I don't think we should hard code else where.

**Pull request merged**